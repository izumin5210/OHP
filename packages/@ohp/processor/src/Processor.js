// @flow
import { bodyToReactFactory, outlineToReactFactory } from './factories'
import type { Options, Result } from './types'

export default class Processor {
  constructor ({ body, outline }: Options = {}) {
    this.bodyToReact = bodyToReactFactory((body ? body.componentByName : null) || {})
    this.outlineToReact = outlineToReactFactory((outline ? outline.componentByName : null) || {})
  }

  bodyToReact: any
  outlineToReact: any

  process = async (body: string): Promise<Result> => {
    const [bodyVfile, outlineVfile] = await Promise.all([
      this.bodyToReact.process(body),
      this.outlineToReact.process(body),
    ])
    return {
      body: bodyVfile.contents,
      outline: outlineVfile.contents,
      styles: bodyVfile.styles,
      meta: bodyVfile.meta,
    }
  }
}
