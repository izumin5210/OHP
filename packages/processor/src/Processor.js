// @flow
import { bodyToReactFactory, outlineToReactFactory } from './factories'
import type { Options, Result } from './types'

export default class Processor {
  constructor (options: Options = {}) {
    this.bodyToReact = bodyToReactFactory(options.componentByName || {})
    this.outlineToReact = outlineToReactFactory(options.componentByName || {})
  }

  bodyToReact: any
  outlineToReact: any

  async process (body: string): Promise<Result> {
    const [bodyVfile, outlineVfile] = await Promise.all([
      this.bodyToReact.process(body),
      this.outlineToReact.process(body),
    ])
    return {
      body: bodyVfile.contents,
      outline: outlineVfile.contents,
      styles: bodyVfile.styles,
      meta: bodyVfile.meta,
    }
  }
}
