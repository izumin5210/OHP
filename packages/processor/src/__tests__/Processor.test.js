// @flow
import React from 'react'
import renderer from 'react-test-renderer'

import { readFileSync as read } from 'fs'
import { join } from 'path'

import Processor from '../Processor'

const FIXTURES = join(__dirname, 'fixtures')

// eslint-disable-next-line react/prop-types
function Page ({ beginLineAt, beginColumnAt, endLineAt, endColumnAt, children, className }) {
  const chunks = []
  if (beginLineAt != null && beginColumnAt != null) {
    chunks.push(`Ln ${beginLineAt}, Col ${beginColumnAt}`)
  }
  chunks.push('~')
  if (endLineAt != null && endColumnAt != null) {
    chunks.push(`Ln ${endLineAt}, Col ${endColumnAt}`)
  }
  return (
    <section {...{ className }}>
      { children }
      <footer>
        { chunks.join(' ') }
      </footer>
    </section>
  )
}

// eslint-disable-next-line react/prop-types
function PageNumber ({ number, enable, className }) {
  if (!enable) {
    return false
  }

  return (
    <span {...{ className }}>
      { number }
    </span>
  )
}

const testcases: Array<{ type: string, options?: any }> = [
  { type: 'simple' },
  { type: 'styles' },
  { type: 'page', options: { body: { componentByName: { page: Page } } } },
  { type: 'pageNumber', options: { body: { componentByName: { pageNumber: PageNumber } } } },
]

testcases.forEach(({ type, options }) => {
  describe(`Processing ${type} document`, () => {
    let result

    beforeEach(async () => {
      const processor = new Processor(options)
      const input = read(join(FIXTURES, `${type}.md`), 'utf-8')
      result = await processor.process(input)
    })

    it('returns react components for body correctly', () => {
      const component = renderer.create(result.body)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('returns react components for outline correctly', () => {
      const component = renderer.create(result.outline)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('returns empty styles', () => {
      expect(result.styles).toMatchSnapshot()
    })
  })
})
