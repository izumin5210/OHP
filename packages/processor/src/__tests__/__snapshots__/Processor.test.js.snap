// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processing page document returns empty styles 1`] = `Array []`;

exports[`Processing page document returns react components for body correctly 1`] = `
<div>
  <section
    className="page"
  >
    <h1>
      Heading 1-1
    </h1>
    <pageNumber
      className=""
      enable="true"
      number="1"
    >
      1
    </pageNumber>
    <footer>
      ~ Ln 8, Col 1
    </footer>
  </section>
  
  
  <section
    className="page toc"
  >
    <h2>
      Heading 2-1
    </h2>
    <ul>
      
      
      <li>
        foo
      </li>
      
      
      <li>
        bar
      </li>
      
      
      <li>
        baz
      </li>
      
      
    </ul>
    <pageNumber
      className=""
      enable="true"
      number="2"
    >
      2
    </pageNumber>
    <footer>
      Ln 8, Col 1 ~ Ln 17, Col 1
    </footer>
  </section>
  
  
  <section
    className="page intro"
  >
    <h2>
      Heading 2-2
    </h2>
    <h3>
      introduction
    </h3>
    <pageNumber
      className=""
      enable="true"
      number="3"
    >
      3
    </pageNumber>
    <footer>
      Ln 17, Col 1 ~
    </footer>
  </section>
</div>
`;

exports[`Processing page document returns react components for outline correctly 1`] = `
<div>
  <ul>
    
    
    <li
      depth="1"
    >
      
      
      <p>
        <a
          href="#heading-1-1"
        >
          Heading 1-1
        </a>
      </p>
      
      
      <ul>
        
        
        <li
          depth="2"
        >
          <a
            href="#heading-2-1"
          >
            Heading 2-1
          </a>
        </li>
        
        
        <li
          depth="2"
        >
          
          
          <p>
            <a
              href="#heading-2-2"
            >
              Heading 2-2
            </a>
          </p>
          
          
          <ul>
            
            
            <li
              depth="3"
            >
              <a
                href="#introduction"
              >
                introduction
              </a>
            </li>
            
            
          </ul>
          
          
        </li>
        
        
      </ul>
      
      
    </li>
    
    
  </ul>
</div>
`;

exports[`Processing pageNumber document returns empty styles 1`] = `Array []`;

exports[`Processing pageNumber document returns react components for body correctly 1`] = `
<div>
  <page
    className=""
    endColumnAt="1"
    endLineAt="10"
  >
    <h1>
      Heading 1-1
    </h1>
  </page>
  
  
  <page
    beginColumnAt="1"
    beginLineAt="10"
    className=""
  >
    <h2>
      Heading 2-2
    </h2>
    <ul>
      
      
      <li>
        foo
      </li>
      
      
      <li>
        bar
      </li>
      
      
      <li>
        baz
      </li>
      
      
    </ul>
    <span
      className="pageNumber intro"
    >
      1
    </span>
  </page>
</div>
`;

exports[`Processing pageNumber document returns react components for outline correctly 1`] = `
<div>
  <ul>
    
    
    <li
      depth="1"
    >
      
      
      <p>
        <a
          href="#heading-1-1"
        >
          Heading 1-1
        </a>
      </p>
      
      
      <ul>
        
        
        <li
          depth="2"
        >
          <a
            href="#heading-2-2"
          >
            Heading 2-2
          </a>
        </li>
        
        
      </ul>
      
      
    </li>
    
    
  </ul>
</div>
`;

exports[`Processing simple document returns empty styles 1`] = `Array []`;

exports[`Processing simple document returns react components for body correctly 1`] = `
<div>
  <page
    className=""
    endColumnAt="1"
    endLineAt="3"
  >
    <h1>
      Heading 1-1
    </h1>
    <pageNumber
      className=""
      enable="true"
      number="1"
    >
      1
    </pageNumber>
  </page>
  
  
  <page
    beginColumnAt="1"
    beginLineAt="3"
    className=""
  >
    <h2>
      Heading 2-2
    </h2>
    <ul>
      
      
      <li>
        foo
      </li>
      
      
      <li>
        bar
      </li>
      
      
      <li>
        baz
      </li>
      
      
    </ul>
    <pageNumber
      className=""
      enable="true"
      number="2"
    >
      2
    </pageNumber>
  </page>
</div>
`;

exports[`Processing simple document returns react components for outline correctly 1`] = `
<div>
  <ul>
    
    
    <li
      depth="1"
    >
      
      
      <p>
        <a
          href="#heading-1-1"
        >
          Heading 1-1
        </a>
      </p>
      
      
      <ul>
        
        
        <li
          depth="2"
        >
          <a
            href="#heading-2-2"
          >
            Heading 2-2
          </a>
        </li>
        
        
      </ul>
      
      
    </li>
    
    
  </ul>
</div>
`;

exports[`Processing styles document returns empty styles 1`] = `
Array [
  "
h1 {
  font-size: 2.6rem;
}
",
  "
li {
  color: #333;
}
",
]
`;

exports[`Processing styles document returns react components for body correctly 1`] = `
<div>
  <page
    className=""
    endColumnAt="1"
    endLineAt="9"
  >
    <h1>
      Heading 1-1
    </h1>
    <pageNumber
      className=""
      enable="true"
      number="1"
    >
      1
    </pageNumber>
  </page>
  
  
  <page
    beginColumnAt="1"
    beginLineAt="9"
    className=""
  >
    <h2>
      Heading 2-2
    </h2>
    <ul>
      
      
      <li>
        foo
      </li>
      
      
      <li>
        bar
      </li>
      
      
      <li>
        baz
      </li>
      
      
    </ul>
    <pageNumber
      className=""
      enable="true"
      number="2"
    >
      2
    </pageNumber>
  </page>
</div>
`;

exports[`Processing styles document returns react components for outline correctly 1`] = `
<div>
  <ul>
    
    
    <li
      depth="1"
    >
      
      
      <p>
        <a
          href="#heading-1-1"
        >
          Heading 1-1
        </a>
      </p>
      
      
      <ul>
        
        
        <li
          depth="2"
        >
          <a
            href="#heading-2-2"
          >
            Heading 2-2
          </a>
        </li>
        
        
      </ul>
      
      
    </li>
    
    
  </ul>
</div>
`;
